src/inausoft.netCLI/Executor.cs

```diff
@@ -1,102 +1,139 @@
-        /// <param name="args">Application's command line arguments.</param>
-        /// <returns></returns>
-        public static int RunCLI(this IServiceProvider serviceProvider, string[] args)
+        public CLFlow UseMapping(Mapping config)
         {
-            var config = serviceProvider.GetService<CLIConfiguration>();
+            _config = config ?? throw new ArgumentNullException(nameof(config));
+            return this;
+        }
 
-            if (config == null)
+        public CLFlow UseServiceProvider(IServiceProvider provider)
+        {
+            _serviceProvider = provider ?? throw new ArgumentOutOfRangeException(nameof(provider));
+            return this;
+        }
+
+        public CLFlow UseDeserializer(IOptionsDeserializer deserializer)
+        {
+            _deserializer = deserializer ?? throw new ArgumentNullException(nameof(deserializer));
+            return this;
+        }
+
+        public int Run(string[] args)
+        {
+            if (_config == null && _serviceProvider == null)
             {
-                throw new InvalidOperationException($"{nameof(CLIConfiguration)} was not registered. Run '{nameof(AddCLI)}' first.");
+                throw new InvalidOperationException($"{nameof(Mapping)} need to be provided either with {nameof(UseMapping)} method or via {nameof(IServiceProvider)}.");
             }
 
+            var config = _config ?? _serviceProvider.GetService<Mapping>() ?? throw new InvalidOperationException();
+
             if (!args.Any())
             {
                 //TODO Add default handling here
-                throw new InvalidCommandException($"Command was not specified.");
+                return Fallback(ErrorCode.UnspecifiedCommand);
             }
 
-            var commandType = config._commandMap.Keys.FirstOrDefault(it =>
-                Attribute.IsDefined(it, typeof(CommandAttribute)) &&
-                (Attribute.GetCustomAttribute(it, typeof(CommandAttribute)) as CommandAttribute).Name == args[0]);
+            var mappingEntry = config.Entries.FirstOrDefault(it =>
+                 Attribute.IsDefined(it.CommandType, typeof(CommandAttribute)) &&
+                 (Attribute.GetCustomAttribute(it.CommandType, typeof(CommandAttribute)) as CommandAttribute).Name == args[0]);
 
-            if (commandType == null)
+            if (mappingEntry == null)
             {
-                throw new InvalidCommandException(args[0], $"Command {args[0]} was not mapped.");
+                return Fallback(ErrorCode.UnrecognizedCommand);
             }
 
-            var command = config.Deserializer.Deserialize(commandType, args.Skip(1).ToArray());
+            object command;
+
+            try
+            {
+                command = _deserializer.Deserialize(mappingEntry.CommandType, args.Skip(1).ToArray());
+            }
+            catch(DeserializationException ex)
+            {
+                return Fallback(ex.ErrorCode.Value);
```

Comment:
If I recall well `DeserializationException` allows null values for `ErrorCode`, this code may fail?

Reply:
Adjusted to not nullable.